src/connection/activate.ts

```diff
@@ -0,0 +1,88 @@
+import { sendAnalyticsEvent } from '@uniswap/analytics'
+import { InterfaceEventName, WalletConnectionResult } from '@uniswap/analytics-events'
+import { Connection } from 'connection/types'
+import { atom } from 'jotai'
+import { useAtomValue, useUpdateAtom } from 'jotai/utils'
+import { useCallback } from 'react'
+import { useAppDispatch } from 'state/hooks'
+import { updateSelectedWallet } from 'state/user/reducer'
+
+import { didUserReject } from './utils'
+
+export enum ActivationStatus {
+  PENDING,
+  ERROR,
+  IDLE,
+}
+
+type ActivationPendingState = { status: ActivationStatus.PENDING; connection: Connection }
+type ActivationErrorState = { status: ActivationStatus.ERROR; connection: Connection; error: any }
+const IDLE_ACTIVATION_STATE = { status: ActivationStatus.IDLE } as const
+type ActivationState = ActivationPendingState | ActivationErrorState | typeof IDLE_ACTIVATION_STATE
+
+const activationStateAtom = atom<ActivationState>(IDLE_ACTIVATION_STATE)
+
+function useTryActivation() {
+  const dispatch = useAppDispatch()
+  const setActivationState = useUpdateAtom(activationStateAtom)
+
+  return useCallback(
+    async (connection: Connection, onSuccess: () => void) => {
+      /*
+       * Skips wallet connection if the connection should override the default
+       * behavior, i.e. install MetaMask or launch Coinbase app
+       */
+      if (connection.overrideActivate?.()) return
+
+      try {
+        setActivationState({ status: ActivationStatus.PENDING, connection })
+        await connection.connector.activate()
+
+        console.debug(`Connection activated: ${connection.getName()}`)
+        dispatch(updateSelectedWallet({ wallet: connection.type }))
+
+        // Clears pending connection state
+        setActivationState(IDLE_ACTIVATION_STATE)
+
+        onSuccess()
+      } catch (error) {
+        // TODO(WEB-3162): re-add special treatment for already-pending injected errors
+        console.debug(`web3-react connection error: ${JSON.stringify(error)}`)
+        console.log(error.message)
```

Comment:
This is overkill, and a `warn` should be used instead:

```suggestion
        console.debug(`Connection failed: ${connection.getName()}`)
        console.warn(error)
```

Reply:
You should not log if the user rejected. Move this logging to _after_ the `didUserReject` check.

Reply:
Ahhh I had added the log temporarily and forgotten to remove. Thanks for catching, I'll start using the `console.log('cartcrom', ...)` notation more now.

As for the existing debug, this is set up right now to help debug that MetaMask loading state issue from a few weeks ago that I've been unable to repro if you remember. Lynn was able to repro it, but it logged `[Object object]` to the console and we were not able to determine what the issue was. Hence the stringify for now. I updated the TODO comment above to note that the logging should be moved under the `didUserReject` check later